File: Trees/94_Binary_Tree_Inorder_Traversal/solution94.cpp
#include <vector>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> output;
        inorder(root, output);
        return output;
    }

private:
    void inorder(TreeNode* root, vector<int>& output) {
        if (root == nullptr) {
            return;
        }
        
        inorder(root->left, output);   // Traverse the left subtree
        output.push_back(root->val);   // Visit the root
        inorder(root->right, output);  // Traverse the right subtree
    }
};

// Time Complexity:
// Space Complexity:

// Runtime1: 0ms, beats 100%
// Memory1: 10.08MB, beats 70.03%

File: Trees/94_Binary_Tree_Inorder_Traversal/solution94.py
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:

        output = []
        def inorder(root):
            if root is None:
                return
            inorder(root.left)
            output.append(root.val)
            inorder(root.right)
        inorder(root)
        return output

# Runtime1: 32ms, beats 75.77%
# Memory1: 16.59MB, beats 21.32%

# Runtime2: 28ms, beats 91.64%
# Memory2: 16.51MB, beats 21.32%

# Runtime3: 33ms, beats 71.29%
# Memory3: 16.62MB, beats 21.32%